src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java

```diff
@@ -247,31 +278,91 @@ private void submitTopicReadOperation() {}
 
   private void submitTopicDeleteOperation() {}
 
+  private int submitBatchRecordRead(final List<O> ops, final int from, final int to) {
+    var submitCount = 0;
+    if (from < to && concurrencyThrottle.tryAcquire()) {
+      try {
+        val consumerConfig = new Properties();
+        val anyOp = ops.get(from);
+        val nodeAddr = anyOp.nodeAddr();
+        val amountOfRecords = to - from;
+        consumerConfig.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, nodeAddr);
+        consumerConfig.put(
+            ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+            "org.apache.kafka.common.serialization.StringDeserializer");
+        consumerConfig.put(
+            ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+            "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        consumerConfig.put(ConsumerConfig.SEND_BUFFER_CONFIG, this.sndBuf);
+        consumerConfig.put(ConsumerConfig.RECEIVE_BUFFER_CONFIG, this.rcvBuf);
+        consumerConfig.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, String.valueOf(amountOfRecords));
+        consumerConfig.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+        final KafkaConsumer<String, byte[]> consumer = new KafkaConsumer<>(consumerConfig);
+        consumer.subscribe(Arrays.asList(anyOp.dstPath())); // Could ops have different dstPaths?
+        List<Integer> sizesOfReadRecords = new ArrayList<>(amountOfRecords);
+        for (var i = from; i < to; i++) {
+          O recordOp = ops.get(i);
+          recordOp.startRequest();
+        }
+        val readRecords = consumer.poll(Duration.ZERO);
+        for (ConsumerRecord<String, byte[]> record : readRecords) {
```

Comment:
1. This will create the new iterator instance, please iterate using an index
2. submitCount should be determined here as the actual count of the returned records
3. for each corresponding operation invoke startResponse and finishResponse here

Reply:
1. It's impossible because we can't use ConsumerRecords<String, byte[]> as List. We can get list from ConsumerRecords by TopicPartition only.